flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/result/ResultStore.java

```diff
@@ -0,0 +1,141 @@
+
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.gateway.service.utils.SqlExecutionException;
+import org.apache.flink.util.CloseableIterator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+/** A result store which stores and buffers results. */
+public class ResultStore {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ResultStore.class);
+
+    private final CloseableIterator<RowData> result;
+    private final List<RowData> recordsBuffer = new ArrayList<>();
+    private final int maxBufferSize;
+
+    private final Object resultLock = new Object();
+    private final AtomicReference<SqlExecutionException> executionException =
+            new AtomicReference<>();
+    private final ResultRetrievalThread retrievalThread = new ResultRetrievalThread();
+
+    public ResultStore(CloseableIterator<RowData> result, int maxBufferSize) {
+        this.result = result;
+        this.maxBufferSize = maxBufferSize;
+        this.retrievalThread.start();
+    }
+
+    public void close() {
+        retrievalThread.isRunning = false;
+        retrievalThread.interrupt();
+
+        try {
+            result.close();
+        } catch (Exception e) {
+            LOG.error("Failed to close the ResultStore. Ignore the error.", e);
+        }
+    }
+
+    public Optional<List<RowData>> retrieveRecords() {
+        synchronized (resultLock) {
+            // retrieval thread is alive return a record if available
+            // but the program must not have failed
+            if (isRetrieving() && executionException.get() == null) {
+                if (recordsBuffer.isEmpty()) {
+                    return Optional.of(Collections.emptyList());
+                } else {
+                    final List<RowData> change = new ArrayList<>(recordsBuffer);
+                    recordsBuffer.clear();
+                    resultLock.notify();
+                    return Optional.of(change);
+                }
+            }
+            // retrieval thread is dead but there is still a record to be delivered
+            else if (!isRetrieving() && !recordsBuffer.isEmpty()) {
+                final List<RowData> change = new ArrayList<>(recordsBuffer);
+                recordsBuffer.clear();
+                return Optional.of(change);
+            }
+            // no results can be returned anymore
+            else {
+                return handleMissingResult();
+            }
+        }
+    }
+
+    private boolean isRetrieving() {
+        return retrievalThread.isRunning;
+    }
+
+    private Optional<List<RowData>> handleMissingResult() {
+        if (executionException.get() != null) {
+            throw executionException.get();
+        }
+
+        // we assume that a bounded job finished
+        return Optional.empty();
```

Comment:
Why does here assume that a bounded job finished?

Reply:
When the retrieval thread is not running and don't get exception, it means we have fetched all the results.